### PR TITLE
[Kernel][Mamba] Optimize Mamba2 SSD prefill Triton kernels

### PR DESCRIPTION
**Pull request by tomeras91**
## Purpose
Optimize Mamba2 SSD prefill Triton kernels through three changes, and clean up dead code.

### 1. Extended autotune search space

Added new Triton configs to `_chunk_scan_fwd_kernel`, `_chunk_state_fwd_kernel`, and `_chunk_cumsum_fwd_kernel`, covering:
- Higher warp counts (num_warps=8) for better memory latency hiding
- Smaller tiles with more software pipelining stages
- Low register pressure configs (num_stages=1) for large `dstate` values (e.g., Codestral-Mamba with dstate=128)

### 2. `fast_exp`: use hardware `exp2` instead of libdevice `exp`

Introduced a shared `fast_exp` Triton helper (`triton_helpers.py`) that computes `exp(x)` as `exp2(x * log2(e))`. `tl.math.exp2` maps directly to a single `ex2.approx.f32` PTX instruction, while `tl.exp` goes through libdevice `__nv_expf` which adds function call overhead and extra range checking. Applied to all SSD kernels (`ssd_chunk_scan`, `ssd_chunk_state`, `ssd_state_passing`) and the Mamba decode kernel (`mamba_ssm`).

### 3. Batch-parallel `_state_passing_fwd_kernel`

Rewrote the state passing kernel to parallelize across sequences. Previously, all chunks from all sequences were serialized in a single loop using `seq_idx` to detect boundaries. Now each sequence gets its own thread block, with chunk ranges derived from the existing `last_chunk_indices` tensor (no new metadata needed). The loop body is branchless (no `seq_idx` load, no comparison, no conditional state reset).

**Practical impact**: prefill typically runs at BS=1 in vLLM's scheduler, so the batch parallelism rarely activates. However the change is never worse at BS=1, gives ~B× improvement for the (rare) BS>1 case, and simplifies the kernel.

### 4. Remove dead `_chunk_state_varlen_kernel`

Removed `_chunk_state_varlen_kernel` from `ssd_chunk_state.py` (~340 lines). This kernel became a dead code path after chunk-aligned mamba (#24683).

### Quality
Run GSM8K on main vs PR:
```bash
lm_eval --model vllm --model_args pretrained=<model> --batch_size auto --trust_remote_code --tasks gsm8k
```

### Performance
Perf bench (using `vllm bench serve`) on main vs PR

### Quality

GSM8K (5-shot, 1×B200) on [nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-BF16](https://huggingface.co/nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-BF16):

| Filter | main | PR |
|---|---|---|
| flexible-extract | 0.5277 ± 0.0138 | 0.5246 ± 0.0138 |
| strict-match | 0.8203 ± 0.0106 | 0.8203 ± 0.0106 |

Results are within error bars — no quality regression.

### Performance

ISL/OSL/BS = 50000/2048/64 on 1×B200 with [nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-BF16](https://huggingface.co/nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-BF16):

<details>
<summary>Full bench command</summary>

```bash
vllm bench serve \
--backend openai \
--host 0.0.0.0 \
--port 8080 \
--model <model> \
--tokenizer <model> \
--num-prompts 192 \
--trust-remote-code \
--ignore-eos \
--num-warmups 20 \
--max-concurrency 64 \
--random-input-len 50000 \
--random-output-len 2048 \
--no-stream \
--dataset-name random \
--save-result
```

</details>

| Metric | main | PR | Δ |
|---|---|---|---|
| Total token throughput (tok/s) | 55,461 | 55,941 | +0.9% |
| Output token throughput (tok/s) | 2,182 | 2,201 | +0.9% |
| Mean TTFT (ms) | 17,224 | 17,027 | -1.1% |
| Median TTFT (ms) | 16,904 | 16,609 | -1.7% |
| Mean TPOT (ms) | 20.67 | 20.49 | -0.9% |
| Mean ITL (ms) | 20.71 | 20.52 | -0.9% |

Kernel-level nsys profiling (median per-invocation latency):

| Kernel | main (μs) | PR (μs) | Δ |
|---|---|---|---|
| `_chunk_scan_fwd_kernel` | 922.8 | 842.4 | **-8.7%** |
| `_chunk_state_fwd_kernel` | 467.1 | 459.7 | -1.6% |
| `_state_passing_fwd_kernel` | 322.0 | 321.4 | -0.2% |
| `_chunk_cumsum_fwd_kernel` | 18.6 | 12.2 | **-34.4%** |
| `_selective_scan_update_kernel` | 18.0 | 16.7 | -7.0% |

The biggest wins are in `_chunk_scan_fwd_kernel` (-8.7%) and `_chunk_cumsum_fwd_kernel` (-34.4%), driven by the expanded autotune configs finding better block sizes for this model's dimensions.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/layers/mamba/ops/ssd_chunk_scan.py [hunk 1]
```diff
@@ -8,13 +8,84 @@
...
 
 TRITON_22 = version.parse(triton.__version__) >= version.parse("2.2.0")
 
 
 @triton.autotune(
     configs=[
+        # =================================================================
+        # Higher warp count configs for better latency hiding
+        # More warps = more instructions in flight = better memory latency hiding
+        # =================================================================
```
Comment (tdoublep) [inline vllm/model_executor/layers/mamba/ops/ssd_chunk_scan.py:22]: Can we quantify the impact of adding these additional configs on the start-up time / time to process first request?

File: vllm/model_executor/layers/mamba/ops/ssd_chunk_state.py [hunk 2]
```diff
@@ -8,13 +8,15 @@
...
 
+from vllm.model_executor.layers.mamba.ops.triton_helpers import fast_exp
 from vllm.triton_utils import tl, triton
 
 from .mamba_ssm import softplus
 
 
 @triton.autotune(
     configs=[
+        triton.Config({"BLOCK_SIZE_H": 1}),
```
Comment (tdoublep) [inline vllm/model_executor/layers/mamba/ops/ssd_chunk_state.py]: This config was actually removed deliberately because it was causing nondeterministic behaviour (see this PR: https://github.com/vllm-project/vllm/pull/25197)

**Review by tdoublep — COMMENTED** [review]
Thank you for this great PR! The kernel change LGTM and is a nice simplification. 

My only concern is around the auto-tune configs and overhead of introducing more. If you want to get this in quick, perhaps we could factor the configs into a separate PR where we quantify the effect on start-up time / time-to-first request?

**Comment by tomeras91** [comment]
> My only concern is around the auto-tune configs and overhead of introducing more.

Adding some measurements of auto-tune overhead. 
<details>
  <summary>benchmarking script</summary>

```python
"""Benchmark autotuning overhead and request latency for Mamba2 models.

Usage:
    # Fresh autotuning (clears cache)
    python benchmark_autotune.py --model nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-BF16

    # With existing cache (skip autotuning)
    python benchmark_autotune.py --model nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-BF16 --use-existing-cache
"""

import argparse
import os
import shutil
import tempfile
import time

import torch
from vllm import LLM, SamplingParams


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--model", type=str, required=True)
    parser.add_argument("--use-existing-cache", action="store_true",
                        help="Use existing Triton cache instead of a fresh one")
    parser.add_argument("--num-warmup", type=int, default=2,
                        help="Number of warmup requests after first request")
    parser.add_argument("--num-bench", type=int, default=5,
                        help="Number of benchmark requests")
    parser.add_argument("--input-len", type=int, default=1024,
                        help="Input prompt length in tokens")
    parser.add_argument("--output-len", type=int, default=128,
                        help="Max output tokens per request")
    args = parser.parse_args()

    cache_dir = None
    if not args.use_existing_cache:
        cache_dir = tempfile.mkdtemp(prefix="triton_bench_")
        os.environ["TRITON_CACHE_DIR"] = cache_dir
        print(f"Using fresh Triton cache: {cache_dir}")
    else:
        print(f"Using existing Triton cache: "
              f"{os.environ.get('TRITON_CACHE_DIR', '~/.triton/cache')}")

    sampling_params = SamplingParams(
        max_tokens=args.output_len, temperature=0.0
    )
    prompt = "Hello " * args.input_len

    # --- Model loading ---
    torch.cuda.synchronize()
    t0 = time.perf_counter()
    llm = LLM(model=args.model, trust_remote_code=True, enable_prefix_caching=False)
    torch.cuda.synchronize()
    load_time = time.perf_counter() - t0
    print(f"\nModel load time: {load_time:.2f}s")

    # --- First request (triggers autotuning) ---
    torch.cuda.synchronize()
    t0 = time.perf_counter()
    llm.generate([prompt], sampling_params)
    torch.cuda.synchronize()
    first_req_time = time.perf_counter() - t0
    print(f"First request (includes autotuning): {first_req_time:.2f}s")

    # --- Warmup ---
    for _ in range(args.num_warmup):
        llm.generate([prompt], sampling_params)
    torch.cuda.synchronize()

    # --- Benchmark ---
    latencies = []
    for i in range(args.num_bench):
        torch.cuda.synchronize()
        t0 = time.perf_counter()
        llm.generate([prompt], sampling_params)
        torch.cuda.synchronize()
        latencies.append(time.perf_counter() - t0)

    avg = sum(latencies) / len(latencies)
    mn = min(latencies)
    mx = max(latencies)

    print(f"\nSteady-state request latency ({args.num_bench} runs):")
    print(f"  Mean:  {avg:.4f}s")
    print(f"  Min:   {mn:.4f}s")
    print(f"  Max:   {mx:.4f}s")

    print(f"\nAutotuning overhead estimate: "
          f"{first_req_time - avg:.2f}s (first request - mean steady-state)")

    if cache_dir:
        print(f"\nTriton cache at: {cache_dir}")
        print("To clean up: "
              f"rm -rf {cache_dir}")

if __name__ == "__main__":
    main()

```
</details>  

The idea is to measure model startup time with a fresh triton cache, measure latency of a single (first) request, and then measure 5 times the latencies of the same request (obviously, while disabling prefix caching). The diff between the first request latency and the latency of subsequent requests is a measure of the autotuning overhead.

Results (single B200, [nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-BF16](https://huggingface.co/nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-BF16)):
* main:
    * First request (includes autotuning): 29.67s
    * Steady-state request latency (5 runs) - mean: 0.4728s
    * Autotuning overhead estimate: 29.20s
* PR:
    * First request (includes autotuning): 39.76s
    * Steady-state request latency (5 runs) - mean: 0.4726s
    * Autotuning overhead estimate: 39.29s

There was no significant change in model startup time itself.
    
**So with a fresh triton cache, the additional configs do add ~10s to first request latency. If a cache with compiled kernels already exists, the autotuning overhead is 10s compared to 7s on main (so only 3s extra).**
IMO The additional warmup overhead is worth the performance benefit.

In the future, I do believe we should refactor mamba config selection and have a similar mechanism to the one in MoE triton path where optimal configs are saved in json files and are loaded on runtime. That would eliminate auto-tuning overhead entirely, and allow for a wider search space for optimal configs.

File: vllm/model_executor/layers/mamba/ops/ssd_chunk_state.py (hunk 2 above)
Comment (tomeras91) [reply vllm/model_executor/layers/mamba/ops/ssd_chunk_state.py]: removed

File: vllm/model_executor/layers/mamba/ops/ssd_chunk_scan.py (hunk 1 above)
Comment (tomeras91) [reply vllm/model_executor/layers/mamba/ops/ssd_chunk_scan.py:22]: Done. See https://github.com/vllm-project/vllm/pull/35397#issuecomment-3981116915

**Review by tdoublep — APPROVED** [review]
LGTM

> In the future, I do believe we should refactor mamba config selection and have a similar mechanism to the one in MoE triton path where optimal configs are saved in json files and are loaded on runtime. 

I agree. At the very least we should have it so that the auto-tuning is done at init time instead of on the first request.
